src/H5Odtype.c

```diff
@@ -332,6 +332,12 @@ H5O__dtype_decode_helper(unsigned *ioflags /*in,out*/, const uint8_t **pp, H5T_t
                     H5MM_xfree(dt->shared->u.compnd.memb);
                     HGOTO_ERROR(H5E_DATATYPE, H5E_CANTDECODE, FAIL, "unable to decode member type")
                 } /* end if */
+                if (temp_type->shared->size == 0) {
+                    for (j = 0; j <= i; j++)
+                        H5MM_xfree(dt->shared->u.compnd.memb[j].name);
```

Comment:
Maybe make a macro or function for freeing incomplete compounds, these loops are repeated in multiple places (and should probably be added to the other errors in the loop as well)

Reply:
In general, especially given the reported failures, it is worth going over this section of code to make sure everything is cleaned up properly on error.